File: Práctica4/java/src/deepspace/GameUniverse.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;
import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author daniel
 */
public class GameUniverse {
    /**
     * Cantidad de medallas necesarias para ganar el juego
     */
    private static final int WIN = 10;
    
    /**
     * Estado del juego
     */
    private GameStateController gameState;
    
    /**
     * Número de turnos
     */
    private int turns;
    
    /**
     * Dado
     */
    private Dice dice;
    
    /**
     * Ïndice de la estación en juego
     */
    private int currentStationIndex;
    
    /**
     * Estación espacial actual
     */
    private SpaceStation currentStation;
    
    /**
     * Array con las estaciones espaciales en juego
     */
    private ArrayList<SpaceStation> spaceStations;
    
    /**
     * Estación enemiga actual
     */
    private EnemyStarShip currentEnemy;
    
    /**
     * Si existe ciudad espacial o no
     */
    private boolean haveSpaceCity;
    
    /**
     * Constructor
     */
    public GameUniverse() {
        turns = 0;
        dice = new Dice();
        spaceStations = new ArrayList<>();
        gameState = new GameStateController();
        currentStationIndex = -1;
        currentStation = null;
        currentEnemy = null;
        spaceStations = new ArrayList<>();
        haveSpaceCity = false;
    }
    
    // Getters
    
    /**
     * Getter del estado de partida
     * @return state
     */
    public GameState getState(){
        return gameState.getState();
    }
    
    /**
     * Descarta un hangar de la estación espacial en juego
     * Solo es posible si el estado del juego es INIT o AFTERCOMBAT
     */
    public void discardHangar(){
        if (GameState.AFTERCOMBAT == gameState.getState() || gameState.getState() == GameState.INIT)
            currentStation.discardHangar();
    }
    
    /**
     * Descarta un arma del hangar de la estación espacial actualmente en juego
     * Solo lo hace si el estado del juego es INIT o AFTERCOMBAT
     * @param i indice del escudo a descartar
     */
    public void discardShieldBoosterInHangar(int i){
        if (GameState.AFTERCOMBAT == gameState.getState() || gameState.getState() == GameState.INIT)
            currentStation.discardShieldBoosterInHangar(i);
    }
    
    /**
     * Descarta un arma del hangar de la estación espacial actualmente en juego
     * Solo lo hace si el estado del juego es INIT o AFTERCOMBAT
     * @param i indice del arma a descartar
     */
    public void discardWeaponInHangar(int i){
        if (GameState.AFTERCOMBAT == gameState.getState() || gameState.getState() == GameState.INIT)
            currentStation.discardWeaponInHangar(i);
    }
    
    /**
     * Elimina un arma montada de la estación espacial actual
     * Solo lo hace si el estado del juego es INIT o AFTERCOMBAT
     * @param i índice del arma a eliminar
     */
    public void discardWeapon(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            currentStation.discardWeapon(i);
        }
    }
    
    /**
     * Elimina un escudo montado de la estación espacial actual
     * Solo lo hace si el estado del juego es INIT o AFTERCOMBAT
     * @param i índice del escudo a eliminar
     */
    public void discardShieldBooster(int i){
        if (getState() == GameState.INIT || getState() == GameState.AFTERCOMBAT){
            currentStation.discardShieldBooster(i);
        }
    }
    
    /**
     * Monta un escudo en la estación espacial actualmente en juego
     * Solo lo hace si el estado del juego es INIT o AFTERCOMBAT
     * @param i indice del escudo a montar
     */
    public void mountShieldBooster(int i){
        if (GameState.AFTERCOMBAT == gameState.getState() || gameState.getState() == GameState.INIT)
            currentStation.mountShieldBooster(i);
    }
    
    /**
     * Monta un arma en la estación espacial actualmente en juego
     * Solo lo hace si el estado del juego es INIT o AFTERCOMBAT
     * @param i indice del arma a montar
     */
    public void mountWeapon(int i){
        if (GameState.AFTERCOMBAT == gameState.getState() || gameState.getState() == GameState.INIT)
            currentStation.mountWeapon(i);
    }
    
    /**
     * Comprueba si la estación espacial jugando en ese turno cumple la condición de victoria
     * @return true en caso afirmativo, false en caso contrario
     */
    public boolean haveAWinner(){
        if (currentStation != null)
            return currentStation.getNMedals() >= WIN;
        else return false;
    }
    
    /**
     * Inicializa una partida
     * @param names array con los nombres de los jugadores
     */
    public void init(ArrayList<String> names){
        if (gameState.getState() == GameState.CANNOTPLAY){
            CardDealer dealer = CardDealer.getInstance();
            Iterator<String> it = names.iterator();
            while(it.hasNext()){
                SuppliesPackage supplies = new SuppliesPackage(dealer.nextSuppliesPackage());
                SpaceStation station = new SpaceStation(it.next(), supplies);
                spaceStations.add(station);
                
                int nh = dice.initWithNHangars();
                int nw = dice.initWithNWeapons();
                int ns = dice.initWithNShields();
                
                Loot lo = new Loot(0,nw,ns,nh,0);
                station.setLoot(lo);
            }
            
            currentStationIndex = dice.whoStarts(names.size());
            currentStation = spaceStations.get(currentStationIndex);
            currentEnemy = dealer.nextEnemy();
            
            gameState.next(turns, spaceStations.size());
        }
    }
    
    /**
     * Si no existe daño pendiente, pasa el turno al siguiente jugador
     * @return true en caso de pasar de turno, false en caso contrario
     */
    public boolean nextTurn(){
        if(gameState.getState() == GameState.AFTERCOMBAT){
            boolean stationState = currentStation.validState();
            if(stationState){
                currentStationIndex = (currentStationIndex+1)%spaceStations.size();
                turns+=1;
                
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedItems();
                
                CardDealer dealer = CardDealer.getInstance();
                currentEnemy = dealer.nextEnemy();
                
                gameState.next(turns, spaceStations.size());
                return true;
            }else{
                return false;
            }
        }else{
            return false;
        }
    }
    
    /**
     * Ejecución del combate
     * @param station estación en combate
     * @param enemy enemigo en cmbate
     * @return resultado del combate
     */
    private CombatResult combat(SpaceStation station, EnemyStarShip enemy){
        GameCharacter ch = dice.firstShot();
        boolean enemyWins;
        CombatResult combatresult;
                
        if (ch == GameCharacter.ENEMYSTARSHIP){
            float fire = enemy.fire();
            ShotResult result = station.receiveShot(fire);
            
            if (result == ShotResult.RESIST){
                fire = station.fire();
                result = enemy.receiveShot(fire);
                
                enemyWins = (result == ShotResult.RESIST);
            }else{
                enemyWins = true;
            }
        }else{
            float fire = station.fire();
            ShotResult result = enemy.receiveShot(fire);
            
            enemyWins = (result == ShotResult.RESIST);
        }
        
        if (enemyWins){
            float s = station.getSpeed();
            boolean moves = dice.spaceStationMoves(s);
            
            if (!moves){
                Damage damage = enemy.getDamage();
                station.setPendingDamage(damage);
                
                combatresult = CombatResult.ENEMYWINS;
            }else{
                station.move();
                
                combatresult = CombatResult.STATIONESCAPES;
            }
        }else{
            Loot aLoot = enemy.getLoot();            
            Transformation transform = station.setLoot(aLoot);
            
            if (transform == Transformation.GETEFFICIENT){
                makeStationEfficient();
                combatresult = CombatResult.STATIONWINSANDCONVERTS;
            } else if (transform == Transformation.SPACECITY){
                createSpaceCity();
                combatresult = CombatResult.STATIONWINSANDCONVERTS;
            } else
                combatresult = CombatResult.STATIONWINS;
        }
        
        gameState.next(turns, spaceStations.size());
        return combatresult;
    }
    
    /**
     * Crea una ciudad espacial
     */
    private void createSpaceCity(){
        if (!haveSpaceCity){
            ArrayList<SpaceStation> rest = new ArrayList<>();
            
            for (SpaceStation station : spaceStations)
                if (station != currentStation)
                    rest.add(station);
            
            currentStation = new SpaceCity(currentStation, rest);
            spaceStations.set(currentStationIndex, currentStation);
            haveSpaceCity = true; 
        }
    }
    
    /**
     * Hace laestación espacial eficiente o eficiente beta
     */
    private void makeStationEfficient(){
        currentStation = new PowerEfficientSpaceStation(currentStation);
        if (dice.extraEfficiency())
            currentStation = new BetaPowerEfficientSpaceStation(currentStation);
        spaceStations.set(currentStationIndex, currentStation);
    }
    
    /**
     * Combate entre una estación espacial y una nave enemiga
     * @return resultado de la pelea
     */
    public CombatResult combat(){
        if (getState() == GameState.BEFORECOMBAT || getState() == GameState.INIT)
            return combat(currentStation, currentEnemy);
        else
            return CombatResult.NOCOMBAT;
    }
    
    /**
     * String representation of the object.
     * @return string representation
     */
    @Override
    public String toString() {
        return  "GameUniverse(\n" +
                "\tcurrentStationIndex = " + currentStationIndex + "\n" +
                "\tcurrentStation = " + currentStation + "\n" +
                "\tcurrentEnemy = " + currentEnemy + "\n" +
                "\tturns = " + turns + "\n" +
                "\tdice = " + dice + "\n" +
                "\tgameState = " + gameState + "\n" +
                "\tspaceStations = " + spaceStations + "\n" +
                "\tWIN = " + WIN + "\n" +
                ")";
    }
    
    /**
     * To UI.
     * @return UI version 
     */
    public GameUniverseToUI getUIversion() {
        return new GameUniverseToUI(currentStation, currentEnemy);
    }
}

File: Práctica4/lib/GameUniverse.rb
#encoding:utf-8

require_relative 'GameUniverseToUI'
require_relative 'GameStateController'
require_relative 'Dice'
require_relative 'Loot'
require_relative 'CombatResult'
require_relative 'GameCharacter'
require_relative 'ShotResult'
require_relative 'SpaceStation'
require_relative 'CardDealer'
require_relative 'EnemyStarShip'
require_relative 'Transformation'
require_relative 'PowerEfficientSpaceStation'
require_relative 'BetaPowerEfficientSpaceStation'
require_relative 'SpaceCity'
module Deepspace
  class GameUniverse

    # @!attribute [Integer] cantidad de medallas necesarias para ganar
    @@WIN = 10

    # Constructor
    def initialize
      # @!attribute [GameStateController] estado de la partida
      @gameState = GameStateController.new

      # @!attribute [Integer] numero de turnos transcurridos
      @turns = 0

      # @!attribute [Dice] game dice
      @dice = Dice.new

      # @!attribute [Integer] indice de la estación espacial en juego
      @currentStationIndex = -1

      # @!attribute [SpaceStation] estación espacial actual
      @currentStation = nil

      # @!attribute [Array<SpaceStation>] array de todas las estaciones espaciales de la partida
      @spaceStations = []

      # @!attribute [EnemyStarShip] enemigo actual
      @currentEnemy = nil

      # @!attribute [Boolean] si existe ciudad espacial o no
      @haveSpaceCity = false
    end

    # Getter for gameState
    # @return [GameState] gameState
    def state
      return @gameState.state
    end

    # Elimina el hangar de la actual estación espacial
    # Solo es posible hacerlo durante la fase de INIT o AFTERCOMBAT
    def discardHangar
      if state == GameState::INIT or GameState::AFTERCOMBAT
        @currentStation.discardHangar
      end
    end

    # Elimina un escudo montado de la estación espacial actual
    # Solo es posible hacerlo durante la fase de INIT o AFTERCOMBAT
    # @param i [Integer] posicion del escudo a eliminar
    def discardShieldBooster(i)
      if state == GameState::INIT or GameState::AFTERCOMBAT
        @currentStation.discardShieldBooster(i)
      end
    end

    # Elimina un escudo almacenado en el hangar de la estación espacial actual
    # Solo es posible hacerlo durante la fase de INIT o AFTERCOMBAT
    # @param i [Integer] posicion del escudo a eliminar
    def discardShieldBoosterInHangar(i)
      if state == GameState::INIT or GameState::AFTERCOMBAT
        @currentStation.discardShieldBoosterInHangar(i)
      end
    end

    # Elimina un arma montada de la estación espacial actual
    # Solo es posible hacerlo durante la fase de INIT o AFTERCOMBAT
    # @param i [Integer] posicion del arma a eliinar
    def discardWeapon(i)
      if state == GameState::INIT or GameState::AFTERCOMBAT
        @currentStation.discardWeapon(i)
      end
    end

    # Elimina un arma almacenada en el hangar de la actual estación espacial
    # Solo es posible hacerlo durante la fase de INIT o AFTERCOMBAT
    # @param i [Integer] posicion del arma a descartar
    def discardWeaponInHangar(i)
      if state == GameState::INIT or GameState::AFTERCOMBAT
        @currentStation.discardWeaponInHangar(i)
      end
    end

    # Monta un escudo del hangar en la estación
    # # Solo es posible hacerlo durante la fase de INIT o AFTERCOMBAT
    # @param i [Integer] posicion en el hangar del escudo a montar
    def mountShieldBooster(i)
      if state == GameState::INIT or GameState::AFTERCOMBAT
        @currentStation.mountShieldBooster(i)
      end
    end

    # Monta un arma del hangar en la estación espacial
    # Solo es posible hacerlo durante la fase de INIT o AFTERCOMBAT
    # @param i [Integer] posicion en el hangar del arma a montar
    def mountWeapon(i)
      if state == GameState::INIT or GameState::AFTERCOMBAT
        @currentStation.mountWeapon(i)
      end
    end

    # Comprueba si la estación espacial actual ha ganado o no
    # @return [Boolean] true en caso afirmativo, false en caso contrario
    def haveAWinner
      if @currentStation.nil?
        raise "Warning! @currentStation nil on GameUniverse.haveAWinner()"
      else
        if @currentStation.nMedals >= @@WIN
          return true
        else
          return false
        end
      end
    end

    # Inicializa un partida
    # @param names [Array<String>] array con los nombres de los jugadores
    def init(names)
      # state = @gameState.state

      if state == GameState::CANNOTPLAY
        dealer = CardDealer.instance
        names.each do |name|
          supplies = dealer.nextSuppliesPackage
          station = SpaceStation.new(name, supplies)
          @spaceStations << station

          nh = @dice.initWithNHangars
          nw = @dice.initWithNWeapons
          ns = @dice.initWithNShields

          lo = Loot.new(0, nw, ns, nh, 0)
          station.setLoot(lo)
        end

        nPlayers = names.length
        @currentStationIndex = @dice.whoStarts(nPlayers)
        @currentStation = @spaceStations[@currentStationIndex]
        @currentEnemy = dealer.nextEnemy

        @gameState.next(@turns, @spaceStations.length)
      end
    end

    # Si no existe daño pendiente, pasa al siguiente jugador
    # @return [Boolean] true en caso de poder pasar de turno, false en caso contrario
    def nextTurn
      # state = @gameState.state

      if state == GameState::AFTERCOMBAT
        stationState = @currentStation.validState

        if stationState
          @currentStationIndex = (@currentStationIndex + 1) \
						             % @spaceStations.length
          @turns += 1

          @currentStation = @spaceStations[@currentStationIndex]
          @currentStation.cleanUpMountedItems

          dealer = CardDealer.instance # behaviour introduced by Singleton
          @currentEnemy = dealer.nextEnemy

          @gameState.next(@turns, @spaceStations.length)

          return true
        end

        return false
      end

      return false
    end

    # Combate entre una estación y una nave enemiga
    # @return [CombatResult] resultado de la batalla
    def combat
      # state = @gameState.state

      if state == GameState::BEFORECOMBAT or GameState::INIT
        return combatGo(@currentStation, @currentEnemy)
      else
        return CombatResult::NOCOMBAT
      end
    end

    # Ejecución del combate
    # @param station [SpaceStation] estación en combate
    # @param enemy [EnemyStarShip] enemigo en combate
    # @return [CombatResult] resultado del combate
    def combatGo(station, enemy)
      ch = @dice.firstShot

      if ch == GameCharacter::ENEMYSTARSHIP
        fire = enemy.fire
        result = station.receiveShot(fire)

        if result == ShotResult::RESIST
          fire = station.fire
          result = enemy.receiveShot(fire)

          enemyWins = (result == ShotResult::RESIST)
        else
          enemyWins = true
        end
      else
        fire = station.fire
        result = enemy.receiveShot(fire)

        enemyWins = (result == ShotResult::RESIST)
      end

      if enemyWins
        s = station.speed
        moves = @dice.spaceStationMoves(s)

        if !moves
          damage = enemy.damage
          station.setPendingDamage(damage)

          combatResult = CombatResult::ENEMYWINS
        else
          station.move

          combatResult = CombatResult::STATIONESCAPES
        end
      else
        aLoot = enemy.loot
        transformation = station.setLoot(aLoot)

        if transformation == Transformation::GETEFFICIENT
          makeStationEfficient
          combatResult = CombatResult::STATIONWINSANDCONVERTS
        elsif transformation == Transformation::SPACECITY
          createSpaceCity
          combatResult = CombatResult::STATIONWINSANDCONVERTS
        else
          combatResult = CombatResult::STATIONWINS
        end
      end

      @gameState.next(@turns, @spaceStations.length)

      return combatResult
    end

    # Crea una ciudad espacial
    def createSpaceCity
      if !@haveSpaceCity
        others = []

        for station in @spaceStations
          if station != @currentStation
            others << station
          end
        end

        @currentStation = SpaceCity.new(@currentStation, others)
        @spaceStations[@currentStationIndex] = @currentStation
        @haveSpaceCity = true
      end
    end

    # Crea una estación espacial eficiente
    def makeStationEfficient
      if @dice.extraEfficiency
        @currentStation = BetaPowerEfficientSpaceStation.new(@currentStation)
      else
        @currentStation = PowerEfficientSpaceStation.new(@currentStation)
      end

      @spaceStations[@currentStationIndex] = @currentStation
    end

    # String representation, UI version
    # ==========================================================================

    # String representation of the object
    # @return [String] string representation
    def to_s
      message = "[GameUniverse] -> Game state: #{@gameState.to_s}," \
				 + "Turns: #{@turns}, Dice: #{@dice.to_s}\n" \
				 + "\tCurrent station: #{@currentStation.to_s}\n" \
				 + "\tCurrent enemy: #{@currentEnemy.to_s}"
      message
    end

    # To UI
    def getUIversion
      return EnemyToUI.new(self)
    end

    # Description:
    # 	Gets the UI representation of the object
    # Returns:
    # 	GameUniverseToUI: the UI representation
    def getUIversion
      return GameUniverseToUI.new(@currentStation, @currentEnemy)
    end
  end

end 
